sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -451,6 +451,11 @@ object ColumnPruning extends Rule[LogicalPlan] {
     case p @ Project(_, g: Generate) if g.join && p.references.subsetOf(g.generatedSet) =>
       p.copy(child = g.copy(join = false))
 
+    // Turn on `omitGeneratorReferences` for Generate if its child column is not used
+    case p @ Project(_, g @ Generate(gu: UnaryExpression, true, _, false, _, _, _))
+      if (AttributeSet(Seq(gu.child)) -- p.references).nonEmpty =>
```

Comment:
@cloud-fan should I also change here from gu.child to gu.references?

Reply:
yea, this should be
```
case p @ Project(_, g @ Generate(generator, join @ true, _, omitGeneratorRefs @ false, _, _, _))
    if p.references.intersect(generator.references).isEmpty =>
  ...
```

Reply:
Actually we can do better, the project may only need part of the generator references. How about we add a new parameter `requiredChildOutputs: Seq[Attribute]` to replace the `join` and `omitGeneratorReferences`?

Reply:
sounds reasonable. let me try that.